### PR TITLE
[spark] Fix insert overwrite with compact error (#2425)

### LINKED ISSUE #2425: [Bug] For primary key table, when use spark insert overwrite some partition and the compact behavior is triggered, IllegalStateException error occurred
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Paimon version

release-0.5

### Compute Engine

Spark

### Minimal reproduce step

1. create table paimon.paimon_test.test_table(id bigint, data string, dt string) partitioned by (dt) TBLPROPERTIES('num-sorted-run.compaction-trigger' = '1', 'primary-key' = 'dt,id');
2. insert into  paimon.paimon_test.test_table values (1, 'test', '2023-11-30');
3. insert overwrite paimon.paimon_test.test_table partition (dt = '2023-11-30') select id,data from paimon.paimon_test.test_table  where dt = '2023-11-30';

### What doesn't meet your expectations?

Exception throw:
Caused by: java.lang.IllegalStateException: Trying to delete file data-44539a97-5060-4b5f-9ea1-9f9119452f50-0.parquet which is not previously added. Manifest might be corrupted.
	at org.apache.paimon.utils.Preconditions.checkState(Preconditions.java:204)
	at org.apache.paimon.manifest.ManifestEntry.assertNoDelete(ManifestEntry.java:154)
	at org.apache.paimon.operation.FileStoreCommitImpl.noConflictsOrFail(FileStoreCommitImpl.java:810)

**the conflict file is the one which added by step two**

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by Xiangakun**
### Purpose
Fix bug: #2425

Linked issue: close #2425

### CODE REVIEW

File: paimon-spark/paimon-spark-3.3/src/test/scala/org/apache/paimon/spark/sql/InsertOverwriteTest.scala [hunk 1]
```diff
@@ -37,7 +37,8 @@ class InsertOverwriteTest extends PaimonSparkTestBase {
 
             spark.sql(s"""
                          |CREATE TABLE T (a INT, b INT, c STRING)
-                         |TBLPROPERTIES ($primaryKeysProp 'bucket'='$bucket')
+                         |TBLPROPERTIES ($primaryKeysProp 'bucket'='$bucket',
+                         |'num-sorted-run.compaction-trigger'='1')
```
Comment (YannByron) [inline paimon-spark/paimon-spark-3.3/src/test/scala/org/apache/paimon/spark/sql/InsertOverwriteTest.scala:41]: Do not use existing test cases to test your cases. Add a new one for it.

**Comment by Zouxxyy** [comment]
@Xiangakun see https://github.com/apache/incubator-paimon/pull/2445

**Comment by Xiangakun** [comment]
And, when use parquet format, seems some jar conflicts:

image

Job aborted due to stage failure: Task 0 in stage 3.0 failed 1 times, most recent failure: Lost task 0.0 in stage 3.0 (TID 3) (10.10.128.57 executor driver): java.lang.NoSuchMethodError: org.apache.parquet.Preconditions.checkArgument(ZLjava/lang/String;Ljava/lang/Object;)V

**Comment by Xiangakun** [comment]
Updated, thanks~

**Comment by Zouxxyy** [comment]
@Xiangakun

> since the writeBuilder in bucketMode match { } is not the same with the one after collect

Can you explain it in more detail?

> And your test case does not cover the real compact action, it will not do a compaction since only one file in each bucket

In fact, the trigger of this bug is that when inserting overwrite partition pk table, restoreFiles + new file can reach `num-sorted-run.compaction-trigger` (no need compaction to be done before), my test can just cover it.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**